New guide: timeline range names (#42853)
mdn/content: files/en-us/web/css/guides/scroll-driven_animations/timeline_range_names/index.md

The animation-range properties accept the keyword `normal`, a {{cssxref("timeline-range-name")}}, a {{cssxref("length-percentage")}}, or both a `<timeline-range-name>` and the `<length-percentage>` we see here. If you wanted the animation to start when 20% of the animation was in view instead of when it's 20% mark was 20% through the viewport, you would set a different `<timeline-range-name>`, which defines the start and end of the range the offsets are relative to.

## Timeline range names

The `<timeline-range-name>` value type accepts six keywords: `cover`, `contain`, `entry`, `exit`, `entry-crossing`, and `exit-crossing`. Each of these represents a predefined named timeline range. A _named timeline range_ is a named segment of an animation timeline. These keywords allow the developer to set the animation attachment range base that offsets are relative to. The start of the segment is represented as `0%` progress through the range; the end of the segment is represented as `100%` progress through the range. Where these points are depend on the named range used.

### Cover

> The animation attachment range in the previous examples all "cover" the entire range. This range represents the full range of the view progress timeline, with `0%` progress representing the point at which the start border edge of the subject aligns with the end edge container and `100%` progress being the when the subject's end border edge reaches the start edge viewport. As we've seen, the size of the `cover` is the sum of subject and viewport dimensions in the scroll direction. In all the examples thus far, the height of the animation attachment range was the height of the container plus the height of the animated element.

Reviewer: ```suggestion
The animation attachment range in the previous examples all "cover" the entire range. This range represents the full range of the view progress timeline. `0%` progress represents the point at which the start border edge of the subject aligns with the end edge of the scrollport, and `100%` progress represents the point where the subject's end border edge reaches the start edge of the scrollport. As we've seen, the size of the `cover` range is the sum of subject and viewport dimensions in the scroll direction. In all the examples thus far, the height of the animation attachment range is the height of the container plus the height of the animated element.
```

    <rect class="large start" width="100" height="500" x="230" y="600" />
    <rect width="96" height="48" x="122" y="602" fill="url(#g)" />
    <rect width="96" height="198" x="232" y="527" fill="url(#g)" />
    <text y="655" x="5">cover</text>
    <text y="590" x="360">60%</text>
    <line x1="0" x2="350" y1="600" y2="600" />
    <line x1="0" x2="350" y1="700" y2="700" />
    <text y="690" x="360">20%</text>
  </svg>
</div>
```

```css hidden live-sample___svg_cover
body div,
body svg {
  margin: -60px 0 -300px;
}
```

{{EmbedLiveSample("svg_cover", "100%", "680")}}

The image demonstrates the animation timeline. The location of the element when it reaches the start of the animation range, either the `0%` view progress or the `20%` view progress mark depending on the `animation-range-start`, is shown in yellow. This represents the position of the element when the `from` keyframe is applied. The red represents the location of the subject element relative to the scrollport when the `to` keyframe is applied. This is the position of the animated element when it reaches the end of the animation. The striped areas in the inset example occur where the red and yellow overlap.

> . T

Reviewer: ```suggestion

```

### Contain

We can set the attachment range to the `contain` named timeline range to have the animation start when 20% of it is 20% of the way through the container and end when 60% of it is 60% of the way through.

The `contain` keyword fully _contains_ the animation within the scrollport, making it so the range starts when the animated element is 100% visible and end when it ceases to be fully visible, if it can be fully visible.

> The `contain` value represents the range during which the principal box is either fully contained by, or fully covers, its view progress visibility range within the scrollport. What the `0%` progress represents depends on whether the animated element is smaller or larger than the scrollport. If the element is smaller than the scrollport in the scroll direction, `0%` occurs when the animated element's end border edge aligns with the end edge of the scrollport and `100%` and `100%` occurs when the animated element's start border edge aligns with the start edge of the scrollport. If the element is larger than the scrollport, the `0%` progress is when the animated element's start border edge reaches the start edge of the viewport and and `100%` is when the end border edge aligns with the end edge. If the animated element is the same size as its container, the animation still happens, but over `0px`.

Reviewer: ```suggestion
The `contain` value represents the range during which the principal box is either fully contained by, or fully covers, its view progress visibility range within the scrollport. What the `0%` progress represents depends on whether the animated element is smaller or larger than the scrollport.

- If the element is smaller than the scrollport in the scroll direction, `0%` occurs when the animated element's end border edge aligns with the end edge of the scrollport, and `100%` occurs when the animated element's start border edge aligns with the start edge of the scrollport.
- If the element is larger than the scrollport, the `0%` progress is when the animated element's start border edge reaches the start edge of the viewport, and `100%` is when the end border edge aligns with the end edge.
- If the animated element is the same size as its container, the animation still happens, but over `0px`.
```

}
```

```css hidden live-sample___inset_contain
body::before {
  content: "contain 20% contain 60%";
  text-align: right;
}
```

```css hidden live-sample___inset_contain
i {
  background-image: linear-gradient(
    to top,
    transparent calc(20% - 1px),
    black calc(20% - 1px) calc(20% + 1px),
    transparent calc(20% + 1px) calc(60% - 1px),
    black calc(60% - 1px) calc(60% + 1px),
    transparent calc(60% + 1px)
  );
}
main {
  background-image: linear-gradient(
    to top,
    transparent calc(20% - 1px),
    black calc(20% - 1px) calc(20% + 1px),
    transparent calc(20% + 1px) calc(40% - 1px),
    lightgrey calc(40% - 1px) calc(40% + 1px),
    transparent calc(40% + 1px) calc(60% - 1px),
    black calc(60% - 1px) calc(60% + 1px),
    transparent calc(60% + 1px) calc(80% - 1px),
    lightgrey calc(80% - 1px) calc(80% + 1px),
    transparent calc(80% + 1px)
  );
  background-origin: content-box;
}
```

{{EmbedLiveSample("inset_contain", "100%", "400")}}

Again we use the `animation-range-start` and `animation-range-end` properties to inset the animation timeline, defining a subsection of the element's full animation attachment range as the active interval.

> If the element is smaller than the scrollport in the scroll direction, the `20%` occurs when the point that is 20% from the animated element's end border edge (which is 60% from its start edge) is 20% of the way from the end edge of the scrollport. The `60%` range end occurs when 20% of the element, which is 60% from the end edge, is 60% of the way through that viewport.

Reviewer: ```suggestion
If the element is smaller than the scrollport in the scroll direction, the `20%` occurs when the point that is 20% from the animated element's end border edge (which is 80% from its start edge) is 20% of the way from the end edge of the scrollport. The `60%` range end occurs when the point 20% from the animated element's start edge, which is 80% from the end edge, is 60% of the way through that viewport.
```

OK, so I get this a bit better than I did during the first draft, but ... WHY does it work like this? Surely it would make more sense if 20% meant that the range start was when the animated element's end edge got to 20% of the way through the scrollport? Mind blown.

## Viewport progress timelines primer

[CSS animations](/en-US/docs/Web/CSS/Guides/Animations) are created by attaching {{cssxref("@keyframes")}} animations to an element using the {{cssxref("animation-name")}} property (or {{cssxref("animation")}} shorthand). The keyframes define the animation's behavior, while the {{cssxref("animation-timeline")}} determines when and how the element progresses through those keyframes.

With [CSS scroll-driven animations](/en-US/docs/Web/CSS/Guides/Scroll-driven_animations/Timelines), the animation's timeline, or progress, is driven by user scroll and element visibility.

In [view-progress timelines](/en-US/docs/Web/CSS/Guides/Scroll-driven_animations/Timelines#view_progress_timelines), keyframe progression is tied to how much of the element is visible within the scroller. As the element enters the viewport, the timeline advances. If the user reverses, the timeline reverses: as the view progress element comes into or moves out of view, the timeline progresses or reverses, respectively. The animation only occurs when the element is visible within its scrollport. If scrolling stops while the element is in view, the animation pauses.

> By default, the view timeline progress starts when the tracked subject's start edge intersects the scrollport at the block or inline end edge and ends when the subject's end edge exits the scrollport at the block or inline start edge.

Reviewer: ```suggestion
By default, the view timeline progress starts when the tracked subject's start edge intersects the scrollport at the end edge and ends when the subject's end edge exits the scrollport at the start edge. If the scrolling occurs in the inline direction, the referenced edges will be the subject and scrollport's inline edges. If the scrolling occurs in the block direction, the referenced edges will be the subject and scrollport's block edges.
```

I'm not sure if this is the most elegant way to say this, but I felt it should be qualified somehow. Alternatively (or in addition), maybe point the reader to https://developer.mozilla.org/en-US/docs/Web/CSS/Guides/Scroll-driven_animations/Timelines#view_progress_timelines for information?

## Setting an animation range

The {{cssxref("animation-range")}} properties define the attachment range for the animation timeline, including the animation attachment range's start and end edges, including insets. The [CSS scroll driven animations module](/en-US/docs/Web/CSS/Guides/Scroll-driven_animations) provides properties and values that accept {{cssxref("length-percentage")}} values (such as `30%` or `100px`) and {{cssxref("timeline-range-name")}} keywords that specify the edges those values are relative to.
The scroll container's writing mode and scroll direction determine the scroll container's start and end edges.

### Setting insets using percentages

> The {{cssxref("animation-range-start")}} and {{cssxref("animation-range-end")}} properties—together forming the `animation-range` shorthand—define an animation's attachment range, limiting the keyframe's active interval to that specific portion of the range.

Reviewer: ```suggestion
The {{cssxref("animation-range-start")}} and {{cssxref("animation-range-end")}} properties—which can both be set using the `animation-range` shorthand—define an animation's attachment range, limiting the keyframe's active interval to that specific portion of the range.
```

```css live-sample___insets
.animated_element {
  animation-range-start: 30%;
  animation-range-end: 70%;
}
```

```css hidden live-sample___insets live-sample___inset_cover live-sample___inset_contain
main,
i {
  background-image: linear-gradient(
    to top,
    transparent calc(30% - 1px),
    black calc(30% - 1px) calc(30% + 1px),
    transparent calc(30% + 1px) calc(70% - 1px),
    black calc(70% - 1px) calc(70% + 1px),
    transparent calc(70% + 1px)
  );
  background-origin: content-box;
}
```

> Here we use `animation-range-start` and `animation-range-end` to inset the animation timeline, defining a subsection of the element's full animation attachment range as the active interval. In this example, the active interval begins `30%` into the default attachment range and ends at `70%` of that same range. Because no `<named-timeline-range>` keyword is included, these values are interpreted relative to the default `cover` animation attachment range.

Reviewer: Explain briefly what `cover` means?

{{EmbedLiveSample("insets", "100%", "400")}}

> For illustrative purposes, horizontal lines were added `30%` from the top and bottom of both the container and the animated element. The percentages are relative to the height of the container and the height of the animated element, not the height of the animation attachment range, which is the sum of the two. Because the full attachment range equals the container height plus the element height, the 30% mark does not occur when the top edge of the animated element aligns with the 30% mark. Instead, it occurs when the top 30% of the element has moved 30% of the way into view.

Reviewer: ```suggestion
For illustrative purposes, horizontal lines are included `30%` from the top and bottom of both the container and the animated element. The percentages are relative to the height of the container and the height of the animated element, not the height of the animation attachment range, which is the sum of the two. Because the full attachment range equals the container height plus the element height, the animation does not start when the top edge of the animated element aligns with the 30% mark. Instead, it occurs when the top 30% of the subject element has moved 30% of the way into view.
```

OK, so I tried testing it in Firefox, and animation ranges don't seem to work there yet (browser compat might be worth a mention somewhere).

I then tried it in Chrome, and ... how this works really is rather confusing. You really would expect `animation-range-start: 30%;` to mean that the animation would start when the subject's top is 30% of the way up the scrollport.

But this is not the case. It doesn't start when the scrollport and subject 30% lines intersect, either. Instead, it starts when the 30% line of the subject is a little way above the 30% line of the scrollport. And then it ends  when the 70% line of the subject is a little way above the 30% line of the scrollport.

As a result, it feels like the 30% lines are somewhat of a red herring, and the line "Instead, it occurs when the top 30% of the element has moved 30% of the way into view." doesn't help me understand what is going on. It should surely be more like "Instead, it occurs when the middle 40% of the element has moved 30% of the way into view." or something? 

I really don't get it, sorry. I think this basic understanding needs more explanation.

### Animation range property values

The animation-range properties accept the keyword `normal`, a {{cssxref("timeline-range-name")}}, a {{cssxref("length-percentage")}}, or both a `<timeline-range-name>` and a `<length-percentage>`.

The `<timeline-range-name>` value type accepts six keywords: `cover`, `contain`, `entry`, `exit`, `entry-crossing`, and `exit-crossing`. Each of these represents a predefined named timeline range. A _named timeline range_ is a named segment of an animation timeline. The start of the segment is represented as `0%` progress through the range; the end of the segment is represented as `100%` progress through the range.

## Cover

> The animation attachment range in the previous examples all "cover" the entire range. We can set the attachment range to a different named timeline range to have the animation start when 30% of it is 30% of the way through the container and end when 70% of it is 70% of the way through. We could have explicitly set a `<timeline-range-name>` to achieve the same results:

Reviewer: I'd expect this section to start with the explanation of what the `cover` keyword means, so, the bit starting "The `cover` value represents the...".

The sentence "We can set the attachment range to a different named timeline range to have the animation start when 30% of it is 30% of the way through the container and end when 70% of it is 70% of the way through." is confusing, as it sounds like you are saying we can set a different keyword other than `cover` to achieve this effect...but isn't this what `cover` does?

The bit about "We could have explicitly set a `<timeline-range-name>` to achieve the same results" I think should be put near the end of the section, after the vital explanation. Maybe start it with "Note that, because `cover` is the default timeline range if one is not named..."

  <rect class="large end" width="100" height="500" x="230" y="225" />
  <rect class="container" width="350" height="250" x="0" y="500" />
  <rect class="small start" width="100" height="50" x="10" y="660" />
  <rect class="medium start" width="100" height="250" x="120" y="600" />
  <rect class="large start" width="100" height="500" x="230" y="525" />
  <rect
    class="medium orange"
    width="96"
    height="48"
    x="122"
    y="602"
    fill="url(#repeat)" />
  <rect
    class="large orange"
    width="96"
    height="198"
    x="232"
    y="527"
    fill="url(#repeat)" />
  <text y="625" x="5">cover</text>
  <text y="590" x="360">70%</text>
  <line x1="0" x2="350" y1="575" y2="575" />
  <line x1="0" x2="350" y1="675" y2="675" />
  <text y="690" x="360">30%</text>
</svg>
```

```css hidden
body {
  display: flex;
  gap: 50px;
}
svg {
  width: 240px;
}

.start,
.end,
.container {
  stroke: black;
  stroke-width: 3;
}
.start {
  fill: yellow;
}
.end {
  fill: red;
}
.container {
  fill: none;
}
text {
  font: 40px monospace;
  fill: black;
}
line {
  stroke: black;
  stroke-width: 6;
  stroke-dasharray: 10;
}
```

{{EmbedLiveSample("cover", "100%", "680")}}

> The image demonstrates the animation timeline. Before the element reaches the start of the animation range, either the `0%` view progress or the `30%` view progress mark depending on the `animation-range-start`, the element is shown in yellow. This represents the position of the element when the `from` keyframe is applied. The red represents the location of the animated element relative to the scrollport when the `to` keyframe is applied. This is the position of the animated element when it reaches the end of the animation. The animation occurs when the element is between these areas, represented by the striped areas.

Reviewer: Some questions about the image:

1. So in each case, the yellow represents the entire subject element in the position just before the animation starts, and the red represents the entire subject element in the position just after the animation ends? And the dashed portion represents the area in which the subject element will be animated? In which case, maybe make the dashed area a different color to the other two, and certainly separate it out from the other two areas? The dashed bit currently looks like it is part of the starting position.
2. Shouldn't all of the size examples have dashed areas? Currently, it is implied that four of the six are not animating, as they don't have dashed areas.
3. This still doesn't really help me understand how the timeline range works. I'm still not sure why the animations start and end when they do, in the case of the 30% and 70%.

## Contain

> To fully _contain_ the animation within the scrollport, making it so the range starts when the animated element is fully visible (if it can be fully visible), we can use `contain`. With `contain`, the `0%` and `100%` both occur when the animated element is visible. The `0%` occurs when the animated element's end edge aligns with the end edge of the scroll container. The `100%` occurs when the start edge of the animated element reaches the start edge of the scroll container. In other words, the `contain` value represents the range of a view progress timeline where the subject element is fully contained by, or fully contains, the view progress visibility range within the scrollport.

Reviewer: ```suggestion
The `contain` keyword fully _contains_ the animation within the scrollport, making it so the range starts when the animated element is 100% visible (if it can be fully visible). With `contain`, the start of the animation (`0%`) occurs when the subject element's end edge aligns with the end edge of the scroll container. The end of the animation (`100%`) occurs when the start edge of the subject element reaches the start edge of the scroll container.
```

  <text y="460" x="10">contain</text>
  <text y="590" x="360">70%</text>
  <line x1="0" x2="350" y1="575" y2="575" />
  <line x1="0" x2="350" y1="675" y2="675" />
  <text y="690" x="360">30%</text>
</svg>
```

```css hidden
body {
  display: flex;
  gap: 50px;
  overflow-y: hidden;
}
svg {
  width: 240px;
  margin: -128px 0;
}
rect {
  stroke: black;
  stroke-width: 3;
}
.start {
  fill: yellow;
}
.end {
  fill: red;
}
.container {
  fill: none;
}
text {
  font: 40px monospace;
  fill: black;
}
line {
  stroke: black;
  stroke-width: 6;
  stroke-dasharray: 10;
}
```

{{EmbedLiveSample("contain", "100%", "400")}}

The animation occurs when the element is between these areas, represented by the white of the container in the `50px` example and by the overlapping orange areas in the `250px` (which is the height of the container) and the `500px` examples.

It may be helpful for some to compare and contrast the `cover` and `contain` values. We can use the shorthand `animation-range` property to declare both the `animation-range-start` and `animation-range-end` have the same `<animation-name-range>` value set at `0%` and `100%` respectively:

```css live-sample___cover_contain
#A {

>   animation-range: contain;g `

Reviewer: ```suggestion
  animation-range: contain;
```

}
#B {
  animation-range: cover;
}
```

We can also use the shorthand to declare offsets:

```css
#A {
  animation-range: contain 30% contain 70%;
}
#B {
  animation-range: cover 30% cover 70%;
}
```

```css hidden live-sample___cover_contain
body::before {
  content: "contain vs cover";
}
.double span {
  display: none;
}
#A,
#B {
  width: 140px;
}
body:has([value="30"]:checked) main,
body:has([value="30"]:checked) #A,
body:has([value="30"]:checked) #B {
  background-image: linear-gradient(
    to top,
    transparent calc(30% - 1px),
    black calc(30% - 1px) calc(30% + 1px),
    transparent calc(30% + 1px) calc(70% - 1px),
    black calc(70% - 1px) calc(70% + 1px),
    transparent calc(70% + 1px)
  );
  background-origin: content-box;
}

#A::after {
  content: " ( contain )";
}
#B::after {
  content: " ( cover ) ";
}
body:has([value="30"]:checked) #A {
  animation-range: contain 30% contain 70%;
}
body:has([value="30"]:checked) #B {
  animation-range: cover 30% cover 70%;
}
```

{{EmbedLiveSample("cover_contain", "100%", "450")}}

> Select different radio buttons and scroll the scrollport to see the different effects of `cover` versus `contain` on animation timelines with shortened ranges.

Author: i added inverted lines for larger than scrollport examples

## Entry and exit

To make the entire animation happen only the subject enters or exits the viewport, use the `entry` or `exit` values respectively. With these two values, the animation attachment range is based on the size of the animated element, not the size of the viewport.

> With `entry`, the `0%` progress occurs the moment the animated element enters the viewport, when the subject's start edge crosses the viewport's end edge. Setting ``animation-range-start: entry 0%` is the same as setting `animation-range-start: cover 0%`.

Reviewer: ```suggestion
With `entry`, the `0%` progress occurs the moment the animated element enters the viewport, when the subject's start edge crosses the viewport's end edge. Setting `animation-range-start: entry 0%` is the same as setting `animation-range-start: cover 0%`.
```

        calc(calc(min(100%, var(--animElHeight)) * 0.3) + 0.5px),
      transparent calc(calc(min(100%, var(--animElHeight)) * 0.3) + 0.5px)
        calc(calc(min(100%, var(--animElHeight)) * 0.7) - 0.5px),
      #f33 calc(calc(min(100%, var(--animElHeight)) * 0.7) - 0.5px)
        calc(calc(min(100%, var(--animElHeight)) * 0.7) + 0.5px),
      transparent calc(calc(min(100%, var(--animElHeight)) * 0.7) + 0.5px)
        calc(min(100%, var(--animElHeight)) - 1px),
      black calc(min(100%, var(--animElHeight)) - 1px)
        calc(min(100%, var(--animElHeight)) + 1px),
      transparent calc(min(100%, var(--animElHeight)) + 1px)
    );
  background-origin: content-box;
}
```

Scroll the viewbox to see the range of the `entry` and `exit` values.

{{EmbedLiveSample("entry_exit", "100%", "500")}}

When the subjects are small enough to be fully contained within the viewport, the animation attachment timeline is at the start (`entry`) or end (`exit`) of the scrollport and the size of the attachment range is limited to the size of the animated element in the scroll direction. The percent offset, if any, is relative to the subject's size, not the viewport size.

> If the animated element is the size of the viewport or larger, the animation doesn't begin until the element fully covers the scrollport in the scroll direction. If your animated element is larger than the scrollport, you may prefer to use `entry-crossing` and `exit-crossing`.

Author: that it is a different article; i only show 30/70 (now 20/60 to make the top and bottom look different) in cover and contain so that going back and forth between the two articles is less necessary during a reading about names.
